scalding-core/src/main/scala/com/twitter/scalding/RichFlowDef.scala

```diff
@@ -146,17 +139,46 @@ class RichFlowDef(val fd: FlowDef) {
     if (sinks.containsKey(pipe.getName)) {
       newFd.addTailSink(pipe, sinks.get(pipe.getName))
     }
+
     // Update the FlowState:
     FlowStateMap.get(fd)
       .foreach { thisFS =>
-        val subFlowState = thisFS.sourceMap
-          .foldLeft(Map[String, Source]()) {
-            case (newfs, kv @ (name, source)) =>
-              if (headNames(name)) newfs + kv
-              else newfs
-          }
-        FlowStateMap.mutate(newFd) { oldFS => (oldFS.copy(sourceMap = subFlowState, flowConfigUpdates = thisFS.flowConfigUpdates ++ oldFS.flowConfigUpdates), ()) }
+        /**
+         * these are all the sources that are upstream
+         * of the pipe in question
+         */
+        val subFlowState =
+          Monoid.sum(
+            thisFS
+              .sourceMap
+              .collect {
+                case (name, source) if headNames(name) =>
+                  FlowState.withSource(name, source)
+              })
+        /*
+         * We assume all the old config updates need to be
+         * done, but this may an over approximation and not
+         * be 100% correct. We have been doing it for a while
+         * however
+         *
+         * Note, this method is only used to convert a pipe
+         * to a TypedPipe. So, we assume there should be
+         * no pending typed writes upstream of this pipe
+         * that are relevant to this pipe when brought
+         * into the TypedAPI
+         */
+        val withConfig = thisFS.copy(sourceMap = Map.empty, pendingTypedWrites = Nil)
+
+        /*
+         * Note that newFd was just allocated, so it has no
+         * FlowState at all, we verify that here to be defensive
+         * since this is not performance critical code
+         */
+        require(FlowStateMap(newFd) == FlowState.empty, s"FlowState is not empty: ${FlowStateMap(newFd)}")
+
+        FlowStateMap.merge(newFd, Monoid.plus(subFlowState, withConfig))
```

Comment:
nit: again confusing name and usage, `FlowState.merge` is an immutable operation, but `FlowStateMap.merge` is mutable, maybe we need use different name or somehow be more clear that we mutate fd here.